caas/kubernetes/provider/application/application.go

```diff
@@ -471,6 +473,81 @@ func (a *app) Ensure(config caas.ApplicationConfig) (err error) {
 	return applier.Run(context.Background(), a.client, false)
 }
 
+// Upgrade upgrades the app to the specified version.
+func (a *app) Upgrade(ver version.Number) error {
+	applier := a.newApplier()
+
+	if err := a.upgradeMainResource(applier, ver); err != nil {
+		return errors.Trace(err)
+	}
+
+	for _, r := range []annotationUpdater{
+		resources.NewSecret(a.secretName(), a.namespace, nil),
```

Comment:
I wonder if we could query the cluster for all resources with the `app.juju.is/created-by` label instead (so longer as the resource also does have the juju version annotation already).
The we don't have to worry about missing anything if something is added later and not also updated here.

Reply:
yeah, that would be ideal. Given we do not have an existing pattern to do this for other similar places(Delete(), and relevant code for podspec applications). I would add a TODO for now.